idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -174,10 +171,35 @@ internal object FirReferenceResolveHelper {
             is FirVariableAssignment -> getSymbolsByVariableAssignment(fir, session, symbolBuilder)
             is FirResolvable -> getSymbolsByResolvable(fir, expression, session, symbolBuilder)
             is FirNamedArgumentExpression -> getSymbolsByNameArgumentExpression(expression, analysisSession, symbolBuilder)
+            is FirValueParameter -> getSymbolForVarargParameterType(fir, session, symbolBuilder)
+                ?: handleUnknownFirElement(expression, analysisSession, session, symbolBuilder)
             else -> handleUnknownFirElement(expression, analysisSession, session, symbolBuilder)
         }
     }
 
+    /**
+     * For a vararg parameter, the corresponding FirElement has a fake source kind
+     * [org.jetbrains.kotlin.fir.FirFakeSourceElementKind.ArrayTypeFromVarargParameter]. In this case, [getOrBuildFir] returns the parent
```

Comment:
`KtReference` is not the only user of `getOrBuildFir`. We can fix the problem for the `getOrBuildFir` in the place where we mark new `FirTypeRef` with `FirFakeSourceElementKind.ArrayTypeFromVarargParameter` by marking delegated type ref with fake source kind instead

Reply:
I am not sure I understand. Consider `foo(vararg i: Int)`. Do you mean to return a `FirTypeRef` of `Int` at line 156? But FIR doesn't contain a type ref for `Int`, it only has a type ref for `IntArray`. Do you mean to somehow create an artificial `Int` type ref and return it in `getOrBuildFir()`?

Reply:
Sorry, if it was not clear. 
I mean fix the way https://github.com/JetBrains/kotlin/blob/721248f8830c9ba9c2862248da929f6ff35dbe79/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/TransformUtils.kt#L98 works: do not mark  created `FirResolvedTypeRef.source` as `FirFakeSourceElementKind.ArrayTypeFromVarargParameter`

Bu we have a contract that two `FirElement`s should not have the same source element. So, we should mark the `FirResolvedTypeRef.delegatedTypeRef.source` with `FirFakeSourceElementKind.ArrayTypeFromVarargParameter` instead

Reply:
Thanks! As part of this change I made `FirUserTypeRefImpl.source` mutable. Is this OK?

Reply:
Actually, it looks like org/jetbrains/kotlin/fir/resolve/transformers/FirSpecificTypeResolverTransformer.kt:71 is violating the contract that two FirElements should not have the same source element. As a result, we have `FirResolvedTypeRef` delegating `FirUserTypeRef` and both have the same source object. Is this intended?